docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+Notification system: a new approach after lot of discussions
+============================================================
+
+Notifications have been a recurrent topic in the last years.
+We have talked about different problems and solution's approaches during these years.
+However, due to the complexity of the change, and without having a clear path, it has been hard to prioritize.
+
+We've written a lot about the problems and potential solutions for the current notification system.
+This is a non-complete list of them just for reference:
+
+* https://github.com/readthedocs/readthedocs.org/issues/4226
+* https://github.com/readthedocs/readthedocs.org/issues/3399
+* https://github.com/readthedocs/meta/discussions/126
+* https://github.com/readthedocs/readthedocs.org/issues/5909
+* https://github.com/readthedocs/readthedocs.org/issues/9110
+* https://github.com/readthedocs/readthedocs.org/issues/9279
+* https://github.com/readthedocs/readthedocs.org/issues/4235
+
+At the offsite in Portland, Anthony and myself were able to talk deeply about this and wrote a bunch of thoughts in a Google Doc.
+We had pretty similar ideas and we thought we were solving most of the problems we identified already.
+
+I read all of these issues and all the discussions I found and wrote this document that summarizes my proposal:
+*create a new notification system that we can customize and expand as we need in the future*:
+
+* A Django model to store the notifications' data
+* API endpoints to retrieve the notifications for a particular resource (User, Build, Project, Organization)
+* Frontend code to display them (*outside the scope of this document*)
+
+
+Goals
+-----
+
+* Keep raising exceptions for errors from the build process
+* Ability to add non-error notifications from the build process
+* Add extra metadata associated to the notification: icon, color, header, body, etc
+* Support different types of notifications (e.g. error, warning, note, tip)
+* Re-use the new notification system for product updates (e.g. new features, deprecated config keys)
+* Message content lives on Python classes that can be translated and formatted with objects (e.g. Build, Project)
+* Message could have richer content (e.g. HTML code) to generate links and emphasis
+* Notifications have trackable state (e.g. unread (default)=never shown, read=shown, dismissed=don't show again)
```

Comment:
I haven't thought about `read_at` and `dismissed_at`, but I think they could be useful. If we implement these fields, we won't need `status` and it will cover all the things I want as well 👍🏼 

* unread: `read_at__isnull=True` (never shown to the user yet)
* read: `read_at__isnull=False` (shown to the user)
* dismissed: `dismissed_at__isnull=False` (the user clicked on the X)

There is an implementation thing that sounds a little weird here, tho. The `read_at` field has to be updated each time the API endpoint is hit to return the notifications. So, it's going to be a `GET` that returns everything but also updates a field?

Reply:
Mm yeah, I was wondering the same thing about the read state update. Whether that is `status=read` or `read_at=...`, we need to update this field at some point around the time that the user receives the message. Our options would be updating on the API response or we could move this to the front end code and update via an API with `status=read`/etc on a UI event -- the user sees the notification or clicks on the notification.

Both are valid, so whatever seems like the best UX. I suspect we'll want the front end version of this though.

Reply:
> we could move this to the front end code and update via an API with `status=read`/etc on a UI event -- the user sees the notification or clicks on the notification.

This is actually the most correct, technically speaking. I'd probably start here, yeah.

Also, I think it's better UX as well since the front-end will get "all the notifications" at once, but may display only 1. In that case, that only one has to be updated with `status=read` 👍🏼 

I will add a PATCH endpoint to update notifications for this use case.